File: JPP/Lista3/zad3/dhsetup.py
from typing import Generic, TypeVar, Callable

import random

T = TypeVar("T")

class DHSetup(Generic[T]):

    def __init__(self, characteristic, field : Callable[..., T]):
        self.modulo = characteristic
        self.field = field
        self.generator : T = self.find_generator()

    def check(self, number):
        tmp = number
        i = 2
        while i * i <= tmp:
            if tmp % i == 0:
                if pow(number, (self.modulo - 1) // i, self.modulo) == 1:
                    return False
                tmp //= i
            else:
                i += 1
        if tmp > 1 and pow(number, (self.modulo - 1) // tmp, self.modulo) == 1:
            return False
        return True

    def find_generator(self) -> T:
        while True:
            g = random.randint(1, self.modulo - 1)
            if self.check(g):
                return self.field(g)

    def power(self, a : T, b) -> T:
        result : T = self.field(1)
        while b > 0:
            if b & 1:
                result = result * a
            a = a * a
            b >>= 1
        return result
    
    def getGenerator(self) -> T:
        return self.generator

File: JPP/Lista3/zad2/DHSetup.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;
import java.util.Random;

class DHSetup<T extends FiniteField> {
    private T a;
    private int modulo;

    private boolean check(long number) {
		long i = 2;
		long tmp = number;

		while (i * i <= tmp) {
			if (tmp % i == 0) {
				if (Math.pow(number, (modulo - 1) / i) == 1) {
					return false;
				}
				tmp /= i;
			} else {
				i++;
			}
		}
		if (tmp > 1 && Math.pow(number, (modulo - 1) / tmp) == 1) {
			return false;
		}
		return true;
	}

    public DHSetup(int characteristic) {
        modulo = characteristic;
        
        Random rand = new Random();
        int newVal;
        T newA;

        do {
            newVal = 1 + rand.nextInt(modulo-1);

        } while (!check(newVal));
        this.a = (T) new FiniteField(newVal);
    }

    public T power(T number, long b) {
		T result = (T) new FiniteField(1);
		while (b > 0) {
			if (b % 2 == 1) {
				result = (T) result.multiply(number);
			}
			number = (T) number.multiply(number);
			b >>= 1;
		}
		return result;
	}

    public T getGenerator() {
        return a;
    }
}
